File: 0_1.baekjoon/binary-search/python/BOJ14225.py
from sys import stdin 
readline = stdin.readline

N = int(readline())
arr = list(map(int, readline().split()))

def solve():
    perm_sum = []
    get_perm_sum(0, 0, perm_sum)

    perm_sum = list(set(perm_sum))
    perm_sum.sort()

    ans = search(perm_sum)
    print(ans)

def search(_sum):
    res = 1
    for num in _sum:
        if res != num:
            return res
        else:
            res += 1
    return _sum[-1]+1

def get_perm_sum(start, cur, res):
    for i in range(start, N):
        res.append(cur+arr[i])
        get_perm_sum(i+1, cur+arr[i], res)

solve()

File: 0_1.baekjoon/binary-search/java/BOJ14225.java
import java.util.*;
import java.io.*;

public class BOJ14225 {
    static int N;
    static int[] arr;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        N = Integer.parseInt(br.readLine());
        
        arr = new int[N];

        StringTokenizer st = new StringTokenizer(br.readLine());
        for(int i=0;i<N;i++){
            arr[i] = Integer.parseInt(st.nextToken());
        }

        solve();
    }
    public static void solve(){
        Set<Integer> permSum = new HashSet<>();
        getPermSum(0, 0, permSum);
        ArrayList<Integer> permSumList = new ArrayList<>(permSum);
        Collections.sort(permSumList);
        int res = 1;
        for(int num : permSumList){
            if (res != num){
                System.out.println(res);
                return;
            } else {
                res++;
            }
        }
        System.out.println(permSumList.get(permSumList.size() - 1) + 1);
    }

    public static void getPermSum(int start, int cur, Set<Integer> res){
        for(int i=start;i<arr.length;i++){
            res.add(cur + arr[i]);
            getPermSum(i+1, cur + arr[i], res);
        }
    }
}
